ddl/placement/rule.go

```diff
@@ -172,24 +172,35 @@ func getYamlMapFormatError(str string) error {
 // NewRules constructs []*Rule from a yaml-compatible representation of
 // 'array' or 'dict' constraints.
 // Refer to https://github.com/pingcap/tidb/blob/master/docs/design/2020-06-24-placement-rules-in-sql.md.
-func NewRules(role PeerRoleType, replicas uint64, cnstr string) ([]*Rule, error) {
-	rules := []*Rule{}
-
+func NewRules(role PeerRoleType, replicas uint64, cnstr string) (rules []*Rule, isDictConst bool, err error) {
 	cnstbytes := []byte(cnstr)
-
 	constraints1, err1 := NewConstraintsFromYaml(cnstbytes)
 	if err1 == nil {
 		rules = append(rules, NewRule(role, replicas, constraints1))
-		return rules, nil
+		err = err1
+		return
+	}
+	// check if is dict constraints
+	constraints2 := map[string]int{}
+	if err2 := yaml.UnmarshalStrict(cnstbytes, &constraints2); err2 != nil {
+		err = fmt.Errorf("%w: should be [constraint1, ...] (error %s), {constraint1: cnt1, ...} (error %s), or any yaml compatible representation", ErrInvalidConstraintsFormat, err1, err2)
+		return
 	}
+	rules, err = NewRulesWithDictConstraints(role, cnstr)
```

Comment:
Here do precheck to ensure that it is in the actual `dict` format, then wrap two errors in `NewRules` or return dict constraints error(definitely it's dict format), it's easier to handle errors. otherwise, we need to unwrap errors from `NewRulesWithDictConstraints` and handle the error to remove the repeated string in the error message.